File: ios/Retry/RetryPolicyEvaluator.swift
//
//  RetryPolicyEvaluator.swift
//  SyncProvider
//
//  Pure-Swift implementation of the retry-backoff math. Mirrors the JS
//  implementation in `src/utils/retryBackoff.ts` 1:1 so behavior is consistent
//  across platforms and the JS-side `SyncOptions.retryPolicy` produces the
//  same delay sequence it would on iOS.
//

import Foundation

/// Sendable Swift mirror of `RetryPolicy` to keep this layer free of the
/// generated `margelo.nitro.syncprovider.RetryPolicy` C++ type. Constructed by
/// the bridge.
struct RetryPolicyValue: Sendable, Equatable {
    let maxAttempts: Int
    let backoff: Backoff
    let baseDelayMs: Int
    let maxDelayMs: Int
    let jitter: Bool
    let retryOnStatusCodes: [Int]

    enum Backoff: String, Sendable {
        case linear = "LINEAR"
        case exponential = "EXPONENTIAL"
        case fibonacci = "FIBONACCI"
    }

    /// Default policy matching the JS facade defaults.
    static let `default` = RetryPolicyValue(
        maxAttempts: 5,
        backoff: .exponential,
        baseDelayMs: 1000,
        maxDelayMs: 60_000,
        jitter: true,
        retryOnStatusCodes: [408, 425, 429, 500, 502, 503, 504]
    )
}

/// Stateless evaluator for ``RetryPolicyValue``.
enum RetryPolicyEvaluator {
    /// Compute the delay (in integer ms) before the next retry, given an
    /// attempt number (1-based — the first retry is `attempt == 1`) and a
    /// policy.
    ///
    /// - Important: **must** match `src/utils/retryBackoff.ts`.
    static func nextDelayMs(attempt: Int, policy: RetryPolicyValue) -> Int {
        let safeAttempt = max(1, attempt)
        let base = max(0, policy.baseDelayMs)

        var raw: Double
        switch policy.backoff {
        case .linear:
            raw = Double(base) * Double(safeAttempt)
        case .exponential:
            raw = Double(base) * pow(2.0, Double(safeAttempt - 1))
        case .fibonacci:
            raw = Double(base) * Double(fibonacci(safeAttempt))
        }

        if policy.jitter {
            let factor = 0.75 + Double.random(in: 0..<0.5)
            raw *= factor
        }

        let cap = policy.maxDelayMs > 0 ? Double(policy.maxDelayMs) : raw
        let clamped = min(max(0, raw), cap)
        return Int(clamped.rounded())
    }

    /// Decide whether an HTTP response should be retried.
    static func shouldRetry(statusCode: Int, policy: RetryPolicyValue) -> Bool {
        return policy.retryOnStatusCodes.contains(statusCode)
    }

    // MARK: - Internals

    private static func fibonacci(_ n: Int) -> Int {
        if n <= 0 { return 0 }
        if n <= 2 { return 1 }
        var prev = 1
        var curr = 1
        for _ in 3...n {
            let next = prev + curr
            prev = curr
            curr = next
        }
        return curr
    }
}

File: android/src/main/java/com/margelo/nitro/syncprovider/retry/RetryPolicyEvaluator.kt
package com.margelo.nitro.syncprovider.retry

import com.margelo.nitro.syncprovider.BackoffStrategy
import com.margelo.nitro.syncprovider.RetryPolicy
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

/**
 * Pure-Kotlin port of the JS-side retry algorithm.
 *
 * The native side and the JS-only fallback MUST agree on the back-off curve
 * for any given `(attempt, policy)` pair, modulo the random jitter component.
 *
 * @see com.margelo.nitro.syncprovider.RetryPolicy
 */
internal object RetryPolicyEvaluator {

  /**
   * Compute the back-off delay, in milliseconds, before [attempt] should
   * be issued. `attempt` is 1-indexed: the first retry uses `attempt = 1`.
   *
   * @param attempt 1-indexed retry attempt number.
   * @param policy active retry policy.
   * @param random PRNG used for jitter; injectable for tests.
   */
  fun computeDelayMs(
    attempt: Int,
    policy: RetryPolicy,
    random: Random = Random.Default,
  ): Long {
    if (attempt < 1) return 0L
    val base = policy.baseDelayMs.coerceAtLeast(0.0)
    val cap = policy.maxDelayMs.coerceAtLeast(base)

    val raw = when (policy.backoff) {
      BackoffStrategy.LINEAR -> base * attempt
      BackoffStrategy.EXPONENTIAL -> base * 2.0.pow((attempt - 1).toDouble())
      BackoffStrategy.FIBONACCI -> base * fibonacci(attempt).toDouble()
    }
    val capped = min(raw, cap)

    val withJitter = if (policy.jitter && capped > 0) {
      // Full jitter [0, capped] (matches the JS-side default behavior).
      capped * random.nextDouble()
    } else {
      capped
    }
    return withJitter.toLong()
  }

  /**
   * Decide whether a request that produced [statusCode] should be retried
   * under [policy]. `null` means the request errored at the transport level
   * (no HTTP response); the caller decides retryability separately.
   */
  fun shouldRetryStatus(statusCode: Int?, policy: RetryPolicy): Boolean {
    if (statusCode == null) return false
    val list = policy.retryOnStatusCodes
    for (i in list.indices) {
      if (list[i].toInt() == statusCode) return true
    }
    return false
  }

  /**
   * Decide whether [attempt] (already incremented past the failed attempt)
   * still has budget under [policy].
   */
  fun hasBudget(attempt: Int, policy: RetryPolicy): Boolean {
    return attempt < policy.maxAttempts.toInt().coerceAtLeast(1)
  }

  private fun fibonacci(n: Int): Long {
    if (n <= 1) return n.toLong()
    var a = 1L
    var b = 1L
    for (i in 3..n) {
      val sum = a + b
      a = b
      b = sum
    }
    return b
  }
}
